### PR TITLE
[EraVM] Adding support of library linkage

### PR DESCRIPTION
**Pull request by PavelKopyl**
# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: lld/include/lld-c/LLDAsLibraryC.h [hunk 1]
```diff
@@ -23,7 +23,9 @@ LLVM_C_EXTERN_C_BEGIN
     Input/output files are transferred via memory buffers. */
 LLVMBool LLVMLinkEraVM(LLVMMemoryBufferRef inBuf, LLVMMemoryBufferRef *outBuf,
                        const char *metadataPtr, uint64_t metadataSize,
-                       char **ErrorMessage);
+                       const char *const *linkerSymbol,
+                       const char linkerSymbolValue[][20],
```
Comment (hedgar2017) [inline lld/include/lld-c/LLDAsLibraryC.h]: Magic number?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

File: lld/include/lld-c/LLDAsLibraryC.h [hunk 2]
```diff
@@ -23,7 +23,9 @@ LLVM_C_EXTERN_C_BEGIN
     Input/output files are transferred via memory buffers. */
 LLVMBool LLVMLinkEraVM(LLVMMemoryBufferRef inBuf, LLVMMemoryBufferRef *outBuf,
                        const char *metadataPtr, uint64_t metadataSize,
-                       char **ErrorMessage);
+                       const char *const *linkerSymbol,
```
Comment (hedgar2017) [inline lld/include/lld-c/LLDAsLibraryC.h]: => `linkerSymbols`

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 1 above)
Comment (hedgar2017) [inline lld/include/lld-c/LLDAsLibraryC.h]: => `linkerSymbolValues`

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 1 above)
Comment (PavelKopyl) [reply lld/include/lld-c/LLDAsLibraryC.h]: Thank you, I've defined it.

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 2 above)
Comment (PavelKopyl) [reply lld/include/lld-c/LLDAsLibraryC.h]: Fixed.

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 1 above)
Comment (PavelKopyl) [reply lld/include/lld-c/LLDAsLibraryC.h]: Thank you, fixed.

File: lld/ELF/Arch/EraVM.cpp [hunk 3]
```diff
@@ -59,6 +59,9 @@ void EraVM::relocate(uint8_t *loc, const Relocation &rel, uint64_t val) const {
   case R_ERAVM_16_SCALE_8:
     add16scaled(val, /*scale=*/8);
     break;
+  case R_ERAVM_32:
+    write32be(loc, static_cast<uint32_t>(val & 0xFFFFFFFF));
```
Comment (akiramenai) [inline lld/ELF/Arch/EraVM.cpp]: I didn't get why and with the mask if you truncate it to 4 bytes anyway.

File: lld/include/lld-c/LLDAsLibraryC.h [hunk 4]
```diff
@@ -19,11 +19,34 @@
 
 LLVM_C_EXTERN_C_BEGIN
 
+#define LINKER_SYMBOL_SIZE 20
```
Comment (akiramenai) [inline lld/include/lld-c/LLDAsLibraryC.h:74]: Could we document what makes us confident that a symbol doesn't exceed this size limit?

File: lld/include/lld-c/LLDAsLibraryC.h [hunk 5]
```diff
@@ -19,11 +19,34 @@
...
+
 /** Performs linkage operation of an object code via lld-as-a-library.
     Input/output files are transferred via memory buffers. */
-LLVMBool LLVMLinkEraVM(LLVMMemoryBufferRef inBuf, LLVMMemoryBufferRef *outBuf,
-                       const char *metadataPtr, uint64_t metadataSize,
-                       char **ErrorMessage);
+LLVMBool LLVMLinkEraVM(LLVMMemoryBufferRef inBuffer,
+                       LLVMMemoryBufferRef *outBuffer,
+                       const char *const *linkerSymbols,
```
Comment (akiramenai) [inline lld/include/lld-c/LLDAsLibraryC.h]: Could we document the parameters? `linkerSymbols` seems to be connected with `linkerSymbolValues` but if there are constraints on where they should point to they aren't obvious from the interface.

File: lld/include/lld-c/LLDAsLibraryC.h [hunk 6]
```diff
@@ -19,11 +19,34 @@
...
+LLVMBool LLVMIsELF(LLVMMemoryBufferRef inBuffer);
+
+/** Returns undefined linker symbol names of the ELF object file passed
+ *  in \p inBuffer. The names are returned via \p linkerSymbols as the
+ *  array of null-terminated strings. \p numLinkerSymbols will point to
+ *  the number of names. For example, if the file has undefined symbols of
+ *  the form "library_path_[0-4]", then this returns just "library_path".
+ *  Caller should dispose the memory allocated for the \p linkerSymbols
+ *  using LLVMDisposeUndefinedSymbolsEraVM */
+void LLVMGetUndefinedSymbolsEraVM(LLVMMemoryBufferRef inBuffer,
```
Comment (akiramenai) [inline lld/include/lld-c/LLDAsLibraryC.h]: Is a triple point necessary here? We return `void` and we allocate memory so could we use the return type instead?

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 7]
```diff
@@ -4,50 +4,135 @@
...
-                  .str();
-  } else {
-    // There is no metadata, but anyway the final size should be the odd number
-    // of words. If the size is already an odd number, we are done.
-    // Otherwise add one zeroed word.
-    padding = ". = ((. >> 5) | 1) << 5;\n";
+/// Returns size of the section \p SectionName in the object file \p file.
+/// Returns zero, if there is no such section.
+static uint64_t getSectionSize(ObjectFile &file, StringRef sectionName) {
+  for (const SectionRef &sec : file.sections()) {
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: Nit: `find` will express the intent more clearly.
Overall it's usually preferable to use algorithms where applicable as when a reader sees an algorithm he can infer much more information straightaway without even reading the body. In this particular case the fact that it's a lookup and the fact that section is supposed to be uniquely named.

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 8]
```diff
@@ -4,50 +4,135 @@
...
+        (symFlags & object::SymbolRef::SF_Undefined)) {
+      StringRef symName = cantFail(sym.getName());
+      if (!symbolsToBeDefined.contains(symName))
+        return true;
+    }
+  }
+  return false;
+}
+
+/// Returns a string with the linker symbol definitions passed in
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp:104]: It explains `LINKER_SYMBOL_SIZE` and the logic of the previous function. I would extract the parts that describe the design part of the solution and put it into the header.

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 5 above)
Comment (akiramenai) [reply lld/include/lld-c/LLDAsLibraryC.h]: After reading the definition: Maybe `linkerSymbolNames` and `linkerSymbolValues`. It makes clear how they are connected.

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 9]
```diff
@@ -4,50 +4,135 @@
...
+              .str();
+      symbolsStr += ";\n";
+    }
+  }
+  return symbolsStr;
+}
+
+/// Returns a linker script when linking executable file.
+static std::string
+creteEraVMLinkerScript(uint64_t metadataSize, const char *const *linkerSymbols,
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: typo

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 10]
```diff
@@ -76,23 +163,51 @@ SECTIONS {                                                                  \n\
...
+    StringRef inData = localInMemBufRefs[0].getBuffer();
+    *outBuffer = LLVMCreateMemoryBufferWithMemoryRangeCopy(
+        inData.data(), inData.size(), "result");
+    return false;
+  }
+
+  uint64_t MDSize = getSectionSize(*static_cast<ObjectFile *>(InBinary.get()),
+                                   ".eravm-metadata");
+
+  std::string linkerScript =
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp:319]: LinkerSymbolDefs isn't a linker script, is it?
Perhaps worth refactoring or documenting what it means and implies if
* we don't have undef symbols
* we have undefs but 0 linkerSymbols
* we have have both undefs and linker symbols.

Otherwise, the logic below seems convoluted.

**Review by akiramenai — CHANGES_REQUESTED** [review]
Could we put some effort into improving readability/maintainability?

**Comment by PavelKopyl** [comment]
> Could we put some effort into improving readability/maintainability?

Thank you very much for your comments! Sure, I'll address them. But I'd start with renaming "linker symbol" -> "library symbol" everywhere, but the intrinsic name. @hedgar2017, @akiramenai, are you OK with this? Original terminology "linker symbol" came from Solidity docs, but it is meaningless in LLVM-based tool chain. Essentially, these are usual symbols (in terminology of ELF format) that are resolved to the library addresses at a linking time. The only difference from the default workflow is that on MC level each library symbol is split into 5 sub-symbols because a library address is a 20-byte relocatable value, but our ELF format is ELF-32 one.

File: lld/ELF/Arch/EraVM.cpp (hunk 3 above)
Comment (PavelKopyl) [reply lld/ELF/Arch/EraVM.cpp]: Sure, that's not needed!

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 4 above)
Comment (PavelKopyl) [reply lld/include/lld-c/LLDAsLibraryC.h:74]: Added the description.

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 5 above)
Comment (PavelKopyl) [reply lld/include/lld-c/LLDAsLibraryC.h]: I agree. I replaced `linkerSymbols`  -> `linkerSymbolNames` and documented them.

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 6 above)
Comment (PavelKopyl) [reply lld/include/lld-c/LLDAsLibraryC.h]: I agree, but let's fix it after all other fixes, because we also need to update the FE part.

File: lld/lld-c/LLDAsLibraryC.cpp (hunk 7 above)
Comment (PavelKopyl) [reply lld/lld-c/LLDAsLibraryC.cpp]: Done.

File: lld/lld-c/LLDAsLibraryC.cpp (hunk 8 above)
Comment (PavelKopyl) [reply lld/lld-c/LLDAsLibraryC.cpp:104]: Done.

File: lld/lld-c/LLDAsLibraryC.cpp (hunk 9 above)
Comment (PavelKopyl) [reply lld/lld-c/LLDAsLibraryC.cpp]: Thanks.

File: lld/lld-c/LLDAsLibraryC.cpp (hunk 10 above)
Comment (PavelKopyl) [reply lld/lld-c/LLDAsLibraryC.cpp:319]: In both cases we generate linker script. In case we are going emit an ELF file, the script is trivial, as it just contains linker symbol definitions. I've added the comments.

**Review by akiramenai — COMMENTED** [review]
Documentation and code improvements look great, thank you.
I agree that library symbols make more sense in terms of terminology. @hedgar2017 is it ok to rename?
Except for that and remained triple pointer, LGTM.

**Comment by hedgar2017** [comment]
> Documentation and code improvements look great, thank you.
> 
> I agree that library symbols make more sense in terms of terminology. @hedgar2017 is it ok to rename?
> 
> Except for that and remained triple pointer, LGTM.

I don't think it makes sense to make linker know about Solidity libraries.

Even Yul is agnostic of libraries and only features `linkersymbol` with an address. In handwritten Yul it is possible to link any address, not just a library.

**Comment by PavelKopyl** [comment]
> Documentation and code improvements look great, thank you. I agree that library symbols make more sense in terms of terminology. @hedgar2017 is it ok to rename? Except for that and remained triple pointer, LGTM.

Thank you for the review!
As for the usage of "library symbol" terminology, after discussing this with @hedgar2017 we decided to keep the "linker symbol".
This seems reasonable because this patch, in fact, implements a generalization of relocatable symbol. In the ELF format the max width of a symbol value is 64 bits. Here we extended it to 160 bits. Yes, currently the only application of this feature is the libraries support, but in future we can employ this for other things.

**Review by akiramenai — APPROVED** [review]
Please squash

**Review by hedgar2017 — APPROVED** [review]

**Comment by hedgar2017** [comment]
You may just remove the lock, as it is not complete and FE is not going to be multi-threaded here anymore.

**Comment by PavelKopyl** [comment]
> You may just remove the lock, as it is not complete and FE is not going to be multi-threaded here anymore.

Yep, will remove it.
